drivers/ieee802154/ieee802154_fsci.c

```diff
@@ -0,0 +1,705 @@
+/*
+ * Copyright (c) 2017 Linaro Ltd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_IEEE802154_DRIVER_LEVEL
+#define SYS_LOG_DOMAIN "ieee802154_fsci"
+#include <logging/sys_log.h>
+
+#include <errno.h>
+
+#include <kernel.h>
+#include <arch/cpu.h>
+
+#include <board.h>
+#include <device.h>
+#include <drivers/console/uart_pipe.h>
+#include <init.h>
+#include <net/net_if.h>
+#include <net/net_pkt.h>
+#include <net/ieee802154_radio.h>
+
+#include "../../subsys/net/ip/l2/ieee802154/ieee802154_frame.h"
+
+#define gFsciIncluded_c			1
+#define gFsciHost_802_15_4_c		1
+#define gMacUsePackedStructs_d		1
+#define gPHY_802_15_4g_d		0
+
+#include "MacTypes.h"
+#include "FsciAspCommands.h"
+#include "FsciCommands.h"
+#include "FsciInterface.h"
+#include "FsciMacCommands.h"
+
+#define FSCI_TX_POWER_MIN		(-19)
+
+enum {
+	FSCI_RX_STATE_SYNC,
+	FSCI_RX_STATE_OG,
+	FSCI_RX_STATE_OC,
+	FSCI_RX_STATE_LEN,
+	FSCI_RX_STATE_PAYLOAD,
+	FSCI_RX_STATE_CRC,
+};
+
+/* Lookup table for TX_POWER */
+static const u8_t tx_power_lt[22] = {
+	2, 2, 2, 2, 2, 2,	/* -19:-14 dBm */
+	4, 4, 4,		/* -13:-11 dBm */
+	6, 6, 6,		/* -10:-8 dBm */
+	8, 8,			/* -7:-6 dBm */
+	10, 10,			/* -5:-4 dBm */
+	12,			/* -3 dBm */
+	14, 14,			/* -2:-1 dBm */
+	18, 18,			/* 0:1 dBm */
+	24			/* 2 dBm */
+};
+
+struct fsci_context {
+	struct net_if *iface;
+	struct k_sem req_sem;
+	struct k_mutex req_mutex;
+	bool req_started;
+
+	u8_t mac_addr[8];
+	u8_t lqi;
+	u16_t pan_id;
+	u8_t channel;
+
+	u8_t pipe_buf[1];
+	u8_t rx_state;
+	u8_t rx_off;
+	u8_t rx_buf[gFsciMaxPayloadLen_c];
+	clientPacket_t fsci_pkt;
+};
+
+static struct fsci_context fsci_context_data;
+
+static void fsci_rx_data(void)
+{
+	struct fsci_context *fsci = &fsci_context_data;
+	mcpsDataInd_t *data = (mcpsDataInd_t *)fsci_context_data.rx_buf;
+	struct net_pkt *pkt = NULL;
+	struct net_buf *frag = NULL;
+	struct ieee802154_address_field *addr_field;
+	struct ieee802154_fcf_seq *fs;
+	u8_t *msdu;
+	u8_t *buf;
+
+	fsci_context_data.lqi = data->mpduLinkQuality;
+
+	pkt = net_pkt_get_reserve_rx(0, K_NO_WAIT);
+	if (!pkt) {
+		SYS_LOG_ERR("No buf available");
+		goto out;
+	}
+
+	frag = net_pkt_get_frag(pkt, K_NO_WAIT);
+	if (!frag) {
+		SYS_LOG_ERR("No frag available");
+		goto out;
+	}
+
+	net_pkt_frag_insert(pkt, frag);
+
+	buf = frag->data;
+	fs = (struct ieee802154_fcf_seq *)buf;
+	memset(fs, 0, sizeof(struct ieee802154_fcf_seq));
+	buf += sizeof(struct ieee802154_fcf_seq);
+
+	fs->fc.frame_type = IEEE802154_FRAME_TYPE_DATA;
+	fs->fc.frame_version = IEEE802154_VERSION_802154_2006;
+	fs->fc.pan_id_comp = 1;
+	fs->sequence = data->dsn;
+
```

Comment:
why do you need to create the fcf/seq header again? That's included in all 15.4 frame, it's never build on reception, just parsed (and that's up to the 15.4 MAC layer to do so, not the driver).